Sources/EmbeddedSwift/Documentation.docc/SDKSupport/Baremetal.md

```diff
@@ -1,3 +1,253 @@
+
+Besides just defining the position of code at runtime, a linker script needs to also:
+- Handle and coordinate the initialization of "zero-fill" global variables (aka the BSS section)
+- Handle and coordinate the initialization of non-zero writable global variables (copying initial values from flash to RAM)
+
+Here's an incomplete sketch of an example linker script:
+
+```
+MEMORY
+{
+   flash (rx)  : ORIGIN = 0x08000000, LENGTH = 1024K
+   sram (rwx)  : ORIGIN = 0x20000000, LENGTH = 320K
+}
+
+SECTIONS
+{
+  .text   : { *(.vectors*) ; *(.text*) } > flash
+  .rodata : { *(.rodata*) ; *(.got*) } > flash
+  .bss    : { *(.bss*) } > sram             ; needs runtime handling
+  .data   : { *(.data*) } > sram AT>flash   ; needs runtime handling
+
+  ...
+}
+```
+
+A sketch of an example corresponding startup code (in C):
+
+```c
+void ResetISR(void) {
+  // Initialize bss section
+  uint8_t *bss = &__bss_start;
+  while (bss < &__bss_end) *bss = 0;
+
+  // Initialize read-write data section
+  extern uint8_t __data_start_flash, __data_start, __data_end;
+  uint8_t *src = &__data_start_flash;
+  uint8_t *dst = &__data_start;
+  while (dst < &__data_end) *dst++ = *src++;
+
+  // Call main
+  extern int main(void);
+  main();
+
+  // If main returns, loop forever
+  while(1);
+}
+```
+
+Both these code snippets are not fully functional, they are only demonstrating the complexity of what the linker script and startup code need to do to initialize handle global variables.
+
+Tip: If this handling is not done correctly, a typical symptom is that global variables "don't work", i.e. reading from them doesn't yield the right value, and writing to them doesn't persist. A good way to double check this is by using a debugging a dumping memory at runtime and checking if it matches the virtual memory layout of the ELF file.
+
+## Vector table and interrupts
+
+The vector table is a critical component that maps hardware interrupts and exceptions to specific handler functions in your code. It's typically placed at the beginning of flash memory and contains function pointers that the processor uses when responding to various events.
+
+The processor automatically jumps to the appropriate handler when an interrupt occurs by indexing into this table. If you don't provide a specific handler, it's common to point to a default handler that can help with debugging.
+
+Example vector table structure:
+
+```c
+// Vector table for ARM Cortex-M
+__attribute__((section(".vectors"))) const void *VectorTable[] = {
+  (void*)0x20008000,  // Initial stack pointer
+  ResetISR,           // Reset handler
+  DefaultHandler,     // NMI handler
+  DefaultHandler,     // Hard fault handler
+  // Additional vectors as needed
+};
+```
+
+If you want to actually handle an interrupt (e.g. a GPIO or UART interrupt) in your Swift code, you can forward declare the function in C, and define it using `@cdecl` in Swift:
```

Comment:
I think using InlineArray here would ease this learning curve by not requiring forward declarations.